Mathlib/Data/Nat/Bitwise.lean

```diff
@@ -44,6 +45,112 @@ namespace Nat
 
 set_option linter.deprecated false
 
+section
+variable {f : Bool → Bool → Bool}
+
+lemma bitwise'_bit' {f : Bool → Bool → Bool} (a : Bool) (m : Nat) (b : Bool) (n : Nat)
+    (ham : m = 0 → a = true) (hbn : n = 0 → b = true) :
+    bitwise' f (bit a m) (bit b n) = bit (f a b) (bitwise' f m n) := by
+  rw [bitwise', binaryRec_eq', binaryRec_eq']
+  · apply Or.inr hbn
+  · apply Or.inr ham
+
+@[simp]
+lemma bitwise_zero_left {m : Nat} : bitwise f 0 m = if f false true then m else 0 := by
+  rfl
+
+@[simp]
+lemma bitwise_zero_right {n : Nat} : bitwise f n 0 = if f true false then n else 0 := by
+  unfold bitwise
+  simp only [ite_self, decide_False, Nat.zero_div, ite_true, ite_eq_right_iff]
+  rintro ⟨⟩
+  split_ifs <;> rfl
+
+@[simp]
+theorem bitwise'_zero_right {m : Nat} :
+    bitwise' f m 0 = bif f true false then m else 0 := by
+  unfold bitwise' binaryRec
+  simp only [Bool.cond_eq_ite, eq_mpr_eq_cast, cast_eq, dite_eq_ite]
+  split_ifs with hx <;> simp only [bit_decomp, binaryRec_zero, hx]
+
+lemma bitwise_zero : bitwise f 0 0 = 0 := by
+  simp only [bitwise_zero_right, ite_self]
+
+@[simp]
+lemma bitwise_succ {n m : Nat} :
```

Comment:
This statement looks pretty awkward to me, especially as a `simp` lemma. I think taking `h : n ≠ 0` would be better?

Reply:
I changed it to be in terms of `succ` exactly because it felt cleaner to have a `simp` lemma without side-conditions, but you are right that the actual statement is not very clean. I'll trust your judgement on which one is better